src/api/spec/models/unregistered_user_spec.rb

```diff
@@ -0,0 +1,124 @@
+
+    it { expect(user.state).to eq('unconfirmed') }
+
+    it { expect(create(:user)).to validate_uniqueness_of(:login).with_message('is the name of an already existing user') }
+  end
+
+  context '#register' do
+    let(:user_attributes) do
+      {
+        realname:              'Tux Penguin',
+        login:                 'tux',
+        password:              'tux123',
+        password_confirmation: 'tux123',
+        email:                 'tux@northpole.org'
+      }
+    end
+    let(:attributes_for_query) { user_attributes.slice(:login, :realname, :email).merge(state: 'confirmed') }
+
+    context 'when not in LDAP mode' do
+      context 'when registration is allowed' do
+        before do
+          allow(::Configuration).to receive(:registration).and_return('allow')
+        end
+
+        subject! { UnregisteredUser.register(user_attributes) }
+
+        it 'creates a new confirmed user' do
+          expect(User.where(attributes_for_query)).to exist
+        end
+      end
+
+      context 'when registration requires confirmation' do
+        let(:attributes_for_query) { user_attributes.slice(:login, :realname, :email).merge(state: 'unconfirmed') }
+
+        before do
+          allow(::Configuration).to receive(:registration).and_return('confirmation')
+        end
+
+        subject { UnregisteredUser.register(user_attributes) }
+
+        it 'throws an exception that confirms the user registration... and creates an unconfirmed user' do
+          expect { subject }.to raise_error(UnregisteredUser::ErrRegisterSave,
+                                            'Thank you for signing up! An admin has to confirm your account now. Please be patient.')
+          expect(User.where(attributes_for_query)).to exist
+        end
+      end
+    end
+
+    context 'when registration is denied' do
+      let(:user_count_before) { User.count }
+
+      before do
+        allow(::Configuration).to receive(:registration).and_return('deny')
+      end
+
+      subject { UnregisteredUser.register(user_attributes) }
+
+      it 'throws an exception' do
+        expect { subject }.to raise_error(UnregisteredUser::ErrRegisterSave, 'Sorry, sign up is disabled')
+      end
+
+      it 'creates no new user' do
+        expect(User.count).to eq user_count_before
+      end
+    end
+
+    context 'in LDAP mode' do
+      before do
+        stub_const('CONFIG', CONFIG.merge('ldap_mode' => :on))
+      end
+
+      context 'when normal user is logged in' do
+        let(:user_count_before) { User.count }
+
+        before do
+          User.current = user
+        end
+
+        subject { UnregisteredUser.register(user_attributes) }
+
+        it 'throws an exception' do
+          expect { subject }.to raise_error(UnregisteredUser::ErrRegisterSave, 'Sorry, new users can only sign up via LDAP')
+        end
+
+        it 'creates no new user' do
+          expect(User.count).to eq user_count_before
```

Comment:
``expect {
subject
}.to not_change`` does not work?

Reply:
No, that does not work because of the exception. But I actually forgot to call subject here... Will fix it.